### PR TITLE
My downloads updates - polling and UI

### LINKED ISSUE #11105: My Downloads page continued improvements and bug fixes
To Do:

- [x] remove download request even if not downloading 
- [x] this computed cleanup 
- [x] no my downloads page if there shouldn't be 
- [x] polling cleanup
- [x] properly handle error messages based on whether or not source device/user device are still on the same network
- [x] download button should be hidden once download intiated

[Slack Message](https://learningequality.slack.com/archives/D01F9GTKDLN/p1692139973072919)

### PR DESCRIPTION
**Pull request by marcellamaki**
Fixes [#11105](https://github.com/learningequality/kolibri/issues/11105)

## Summary
This PR addresses some linger UI/UX issues with the My Downloads page.

-  Allows user to remove a download request even if not yet downloaded (i.e. waiting, in progress)
-  Addresses lingering PR feedback for using computed props
-  Prevents access to My Downloads page for users where downloading is not enabled
-  Adds polling for the download status, as well as the network devices
-  Properly handles error messages based on whether or not source device/user device are still on the same network
-  Minorly refactors `useDownloads` composable so that download button is be hidden once download is initiated

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: This seems like an error - `this` does not refer to the component inside the `setup` context, as it is executed before the component VM is created.

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: I think my suggestion was to turn this into a computed prop, but this is still just a plain function.

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: This is a function, not a computed prop.

Also, if you want to do this mapping, you can just map in the object that is the return value from `setup`:

```
downloads: sortedFilteredDownloads,
```

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue:451]: No need to change this, as it was like this when you got here - but could also use the early return pattern here in this switch statement.

File: kolibri/plugins/learn/assets/src/my_downloads/views/MyDownloads.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/MyDownloads.vue]: Every time you poll, your list of devices is going to get longer (with duplicates).

I think we could consolidate some of this logic somewhat across the different places that the learn `useDevices` composable is used, but might be good for us to cohack on that, as I've made some tweaks to this in the place that I suspect you copied this from.

**Review by rtibbles — CHANGES_REQUESTED** [review]
A couple of things that look like they will be clear bugs, and some cleanup that can happen after merge.

File: kolibri/plugins/learn/assets/src/composables/useSearch.js
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/composables/useSearch.js]: As they are in the same folder, you don't _have_ to change the import path here :)

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: Should this be a length check on the Object.keys as well? The object should always be truthy.

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [reply kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: This function is still being invoked in two places in the component, but there's absolutely no need to, as you now have downloads as a computed property for the return value of this.

Could remove the return from `setup` remove the references, and then update this line to:
```suggestion
      const downloads = computed(() => {
```
(and add the matching closing parenthesis at the end, and remove the other definition of `downloads` below).

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue:351]: Your downloads computed prop always returns an array now, so I think you can do without the initial existence check (but no need to remove it).

File: kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue]: This shouldn't be necessary to access the value in a template. If you are having to do this, something has gone awry.

File: kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue]: I think we should set this to start as `true` for initial page load, then it just gets set false at the first time we have a reason to stop showing the loading state.

File: kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue:128]: So, here's the issue - we need to read this from the `useDevices` composable's `networkDevices` otherwise all the work you've done there to ensure that devices that are no longer returned from the API get evicted from our map will have gone to waste, because our `deviceChannelsMap` is never evicting anything.

So this should be reverted, the `deviceChannelsMap` should be `[instance_id]: [<channel_1>, <channel_2>....]` and the filter below should be:

```
.filter(device => this.deviceChannelsMap[device.instance_id]?.length > 0)
```

File: kolibri/plugins/learn/assets/src/views/HybridLearningContentCard/HybridLearningFooter.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/HybridLearningContentCard/HybridLearningFooter.vue]: I don't understand why this is preferable to keying the downloadRequestMap by contentnode_id?

We should revert this change, and instead ensure that the downloadRequestMap is always keyed by contentnode_id.

File: kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js:116]: This function is still assuming that the downloadRequestMap is keyed by contentnode_id.

**Review by rtibbles — CHANGES_REQUESTED** [review]
downloadRequestMap should be keyed by contentnode_id.

Final cleanup of the device channels loading will fix the never purging problem, but it's on the right track.

File: kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js]:
```suggestion
  const downloads = computed(sortedFilteredDownloads);
```

No `this`, and shouldn't need the additional arrow function wrapping.

File: kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/composables/useDownloadRequests.js:41]: I can't comment on the actual line, but please update the setting of the downloadRequestMap to `obj.contentnode_id` not `obj.id`.

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: Isn't this duplicative of what's in the composable now?

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: This is no longer defined - I think this is just `this.downloads` now.

Also, this prop confuses me, because I am not sure what it is calculating.

**Review by rtibbles — DISMISSED** [review]
Still a bit confused by some aspects of the download display.

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: log!

File: kolibri/plugins/learn/assets/src/my_downloads/views/MyDownloads.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/MyDownloads.vue:103]: Is fetchDownloads now only called after a 1 second delay after `created` because we start polling?

File: kolibri/plugins/learn/assets/src/my_downloads/views/MyDownloads.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/MyDownloads.vue:115]: Might be good to do `this.fetchDownloads().then(this.startPolling)`, so that we don't delay the initial fetch by 1 second.

**Review by rtibbles — COMMENTED** [review]
A couple more things, and we've chatted on slack about some additional tweaks - coercing topicsMore to a boolean, and possible updates to the design page.

One small loading thing could be improved on Explore Libraries - it seems we never turn off the core loading state, and it would be nice to use a circular loader while loading is true.

File: kolibri/plugins/learn/assets/src/routes/index.js
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/routes/index.js]: Maybe do this after the logged in guard below?

File: kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue:21]: This is a horrendously confusing diff, but I assume this is just adding an if else?

**Review by rtibbles — COMMENTED** [review]
One more thing, one question. Have not done final manual confirmation yet!

File: kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue
Comment (marcellamaki) [reply kolibri/plugins/learn/assets/src/views/ExploreLibrariesPage/index.vue:21]: yes

**Comment by marcellamaki** [comment]
@radinamatic @pcenov - Richard is doing some manual QA, but this will be ready for the QA team as well! 

So, the things to confirm here would be testing the gherkins for the MyDownloads page, and specific changes here include: 
- A user should be able to remove a download request even if not yet downloaded (i.e. waiting, in progress)
- My Downloads page is not accessible for users where downloading is not enabled (this would be at  
- Properly handles error messages based on whether or not source device/user device are still on the same network (this should be aligned with the figma/gherkin) 
- Updates figma so there is a "status" column, and uses "on your device" message when download has successfully completed

**Comment by radinamatic** [comment]
First impressions, not full walk through all the workflows:

1. As reported on Slack, the bottom bar on Android is a _blocker_:

   
Screenshot_2023-09-29-23-01-17-502_org learningequality Kolibri


2. 'My downloads' is not appearing immediately in the sidebar menu after saving changes in Device settings to enable it: only after the user navigated somewhere (Library/Home), and only then re-opened the sidebar menu
3. First set of resources selected to download remain stuck in the permanent _Waiting to download_ state (LoD being on same WiFi as the server device on Ubuntu 20.04). User on LoD can see the channels on the server, browse its library, and can import resources from it, but the actual download does not start not even after several Kolibri app restarts. What did the trick at the end was rebooting the LoD device (similar as for syncing 🤔 ).
4. Filtering downloads by 'Newest' and 'Oldest' is not working well, filtering by file size looks Ok. 

[LoD-logs-Android13.zip](https://github.com/learningequality/kolibri/files/12772137/LoD-logs-Android13.zip)
[ServerLogs-Ubuntu20.04.zip](https://github.com/learningequality/kolibri/files/12772138/ServerLogs-Ubuntu20.04.zip)

**Comment by pcenov** [comment]
Hi @marcellamaki and @radinamatic, so happy to see this one almost ready now!
I confirm all of Radina's observations and also confirm that the following is implemented as specified:
- users are able to remove a download request even if not yet downloaded
- My Downloads page is not accessible for users if downloading is not enabled
- there is a "status" column and "On your device" message

I'm still testing edge cases about the download statuses.

Other than that I've identified the following issues which I am listing here for now:

| Issue| Screenshot |
:-------------------------:|:-------------------------:
| Console error on page load  |  |
:-------------------------:|:-------------------------:
|The Remove button is not fully visible even if the screen is fully expanded and there are horizontal scrollbars for View/Remove in languages other than English and when in resolutions less than 990px |  |
:-------------------------:|:-------------------------:
| When I click on 'Remove' there's no confirmation modal (which is displayed if I attempt to remove several resources)  | https://github.com/learningequality/kolibri/assets/79847249/6d2a23d5-751a-4326-ab31-d65092d59a63 |
:-------------------------:|:-------------------------:
| Similar to the previous one, but once I delete the last available resource on a page I should be brought back to the previous page which is not happening at the moment  | https://github.com/learningequality/kolibri/assets/79847249/76d79510-d395-4408-aad0-52d41a342740 |
:-------------------------:|:-------------------------:

**Comment by pcenov** [comment]
While testing the statuses I noticed that if I keep connecting/disconnecting the device from which I am downloading the resources, there's some filtering going on which results in disappearance of resources which are supposedly downloaded. I see a `TypeError: this.networkDevices.filter is not a function` error in the console:

Here's a video of that:

https://github.com/learningequality/kolibri/assets/79847249/1d2b419c-0dd1-437c-b6af-14a0da38c0a7

It's also worth mentioning that if I download a resources and then delete it, when I download it again from the other device it shows incorrect 'Date added' value which is the same as the first time the resource was added:

https://github.com/learningequality/kolibri/assets/79847249/04c807c8-d09f-4e80-9ca6-10fde864ff2c

**Comment by marcellamaki** [comment]
Thank you @pcenov for the feedback! I've been working on fixes for yesterday's comments, and will tackle the new items you added today.

**Comment by marcellamaki** [comment]
@pcenov and @radinamatic - when testing the "remove resource" behavior, did you notice anything with your machine _following_ the removal of the device such as higher CPU usage, slower responsiveness of Kolibri, etc.? Alternatively, some very very very spammy logs would be an indicator of something gone awry. I did some troubleshooting with @rtibbles on this earlier this week, but now in my own testing for fixing some of these bugs, I'm definitely noticing some strange behavior, and I'm trying to identify if it's replicable.

**Comment by pcenov** [comment]
@marcellamaki I've just tested the removal of resources from 'My downloads' while monitoring the CPU usage and I am not observing anything abnormal. However since I was testing completely offline, I realized that all of the resources which I removed from 'My downloads' are still available in Device > Channels and 'Your library' section of the 'Library' and are actually still downloaded on the device. So when we remove a resource from 'My downloads' we should make sure these get removed completely.

**Comment by marcellamaki** [comment]
Fixes: 
- console log error 
- remove button visibility/spacing on smaller screens 
- "remove" confirmation modal on single item
- newest/oldest filtering
- "no downloaded resources" message displayed properly when all resources have been deleted 
- TypeError: this.networkDevices.filter is not a function error in the console

> all of the resources which I removed from 'My downloads' are still available in Device > Channels and 'Your library' section of the 'Library' and are actually still downloaded on the device. So when we remove a resource from 'My downloads' we should make sure these get removed completely.

This is currently the way that the downloading / deleting behavior was implemented, so this can be resolved in a follow up, I think 

Follow ups would be: 
- bottom bar alignment (needs design feedback) 
- changes to delete behavior as described above (if any - needs final decision) 
- "my downloads" appearing in sidebar (in my opinion, a big _maybe_)

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue:93]: There's probably a way we could nest the conditionality slightly more to reduce the repetition here, but let's not change it now.

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: We are triggering this method when pressing the 'remove' button on a single download, but this doesn't actually seem to be doing anything, except adding it to the selection list - so this seems like the remove button is only doing the same as selecting the item now?

Seems like this should actually be:
```
this.$emit('removeResources', [download]);
```

If this was needed to get rid of the confirmation dialog, we might need two events, one for `removeResourcesWithConfirmation` which the `removeResources` method uses, and one for `removeResourcesWithoutConfirmation` that is used here.

**Review by rtibbles — COMMENTED** [review]
Just one more thing...

File: kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue
Comment (marcellamaki) [reply kolibri/plugins/learn/assets/src/my_downloads/views/DownloadsList/index.vue]: Updated to rename functions for more clarity around their purposes

**Review by rtibbles — COMMENTED** [review]
All good on the code front, final QA approval and this should be good to go with the follow up issues filed.

**Comment by radinamatic** [comment]
Good news is that I can confirm the fix of all the bullet points listed by @marcellamaki, except the console errors that I could not observe directly while testing on 2 Android devices.

On other hand there are 2 blocking issues present:
1. Users are able to download resources (and go to the **My downloads** page) by default, immediately after the setup wizard, even without enabling the corresponding option in the device settings:

    https://github.com/learningequality/kolibri/assets/1457929/40889e29-42a2-4200-a3df-0d04c0c014dc

2. Syncing is not working at all, not even with restarting apps, rebooting the LoD devices, or stopping/restarting the server with the full facility. The only time the *Synced* status appeared was after the initial user import/creation, and after that devices permanently displayed *Not connected to server* status. 



> This is currently the way that the downloading / deleting behavior was implemented

I'm not sure I understand the point of this implementation... 😕 
What is the point of having the `Remove` button if not to completely remove downloaded resources from the device? If deleting the resources they downloaded directly is the only way learners have to manage and free some space on their device, it makes no sense to me not to actually delete them. 

We may want to consider bringing the **Save changes** button in focus (and the page would scroll down) after the user enables something in Device settings, because it's easy to miss that step on narrow screens where the button is all the way down 😅 
I'll open a separate issue to discuss UX implications.

[Android11-tablet-logs.zip](https://github.com/learningequality/kolibri/files/12826709/Android11-tablet-logs.zip)
[Android13-phone-logs.zip](https://github.com/learningequality/kolibri/files/12826710/Android13-phone-logs.zip)
[Ubuntu20-server-logs.zip](https://github.com/learningequality/kolibri/files/12826711/Ubuntu20-server-logs.zip)

**Comment by pcenov** [comment]
I confirm Radina's observations and here are just a few more that I spotted:

| Issue| Screenshot |
:-------------------------:|:-------------------------:
| This one is still not fixed - Once I delete the last available resource on a page I should be brought back to the previous page/first page  | https://github.com/learningequality/kolibri/assets/79847249/76d79510-d395-4408-aad0-52d41a342740 |
:-------------------------:|:-------------------------:
| If I delete all downloaded channels on the device and then I go to My downloads I can still see the downloaded resources as being 'On your device' and attempting to view any of them results in an error page | https://github.com/learningequality/kolibri/assets/79847249/447b6bdf-83be-45ed-97bc-51d2c4cb4846 |
:-------------------------:|:-------------------------:
| `'TypeError: this.$nextTick.then is not a function'` error after I remove a resource |  |
:-------------------------:|:-------------------------:

**Review by rtibbles — COMMENTED** [review]
Code changes look correct to me - just need to tweak the tests and write follow up issues.

**Comment by pcenov** [comment]
Hi @marcellamaki the only new issue relevant to the latest changes here is that when I have selected all items on the first page and I go to the second page, the 'Select all' checkbox is marked as selected on the second page while it shouldn't be. Also if I attempt to select all items on the second page I have to click the box several times for it to start working and then if I move to the 3-rd page it stops working altogether:

https://github.com/learningequality/kolibri/assets/79847249/4de46593-e3a5-4c09-bbcb-6dcb70306575

I also noticed a potentially confusing behavior which I mentioning as something to be considered: the value for the 'Available storage' doesn't change depending on the limitation set in the Settings, so that if I have not set a download limit it will say 21 GB and if I have set a limit of 1 GB it will still show 21 GB although technically that's not the available storage for downloads.

**Comment by radinamatic** [comment]
Retested with the newest assets, syncing is working great with non-Android LoD.
Let's make sure we collect and report everything pending from this PR for the follow-ups, but otherwise good to go! 👍🏽

**Review by radinamatic — APPROVED** [review]
This is good to go, pending items to be addressed as follow-ups 💯 :shipit: 

Thank you @marcellamaki!

**Review by rtibbles — APPROVED** [review]
I tried to find one final nit to pick, but I could not.

**Pull request merged**
